src/CohortManager.js

```diff
@@ -0,0 +1,114 @@
+  }
+
+  createCohort(cohortName) {
+    const cohortFound = this.cohortList.find((item) => item.name === cohortName)
+    if (cohortFound) {
+      throw new Error('Cohort name is already in use')
+    } else {
+      this.newCohortID++
+      const newCohort = new Cohort(cohortName, this.newCohortID)
+      this.cohortList.push(newCohort)
+    }
+  }
+
+  findCohort(cohortName) {
+    const cohortFound = this.cohortList.find(
+      (item) => item.cohortName === cohortName
+    )
+    if (cohortFound) {
+      return cohortFound
+    } else {
+      throw new Error('Cohort not found')
+    }
+  }
+
+  addStudent(firstName, lastName, gitName, email, cohortName) {
+    const definedParameters = Object.keys([
+      firstName,
+      lastName,
+      gitName,
+      email,
+      cohortName
+    ])
+    const cohortFound = this.cohortList.find(
+      (item) => item.cohortName === cohortName
+    )
+    if (definedParameters.length !== 5) {
+      throw new Error('Please make sure to provide all student details')
+    } else if (!cohortFound) {
+      throw new Error('Cohort not found')
+    } else if (cohortFound.cohortStudents.length >= 24) {
+      throw new Error('Cohort is full')
+    } else {
+      this.newStudentID++
+      const student = new Student(
+        this.newStudentID,
+        firstName,
+        lastName,
+        gitName,
+        email
+      )
+      cohortFound.cohortStudents.push(student)
+    }
+  }
+
+  removeCohort(cohortName) {
+    const cohortFound = this.cohortList.find(
+      (item) => item.cohortName === cohortName
+    )
+    if (cohortFound) {
+      const deletionIndex = this.cohortList.indexOf(cohortFound)
+      this.cohortList.splice(deletionIndex, 1)
+    } else {
+      throw new Error('Cohort not found')
+    }
+  }
+
+  removeStudent(cohortName, studentFirstName, studentLastName) {
+    const cohortFound = this.cohortList.find(
+      (item) => item.cohortName === cohortName
+    )
+    if (!cohortFound) {
+      throw new Error('Cohort not found')
+    }
+    const studentFound = cohortFound.cohortStudents.find(
+      (item) =>
+        item.firstName === studentFirstName && item.lastName === studentLastName
+    )
+    if (studentFound) {
+      const deletionIndex = cohortFound.cohortStudents.indexOf(studentFound)
+      cohortFound.cohortStudents.splice(deletionIndex, 1)
+    } else {
+      throw new Error('Student not found')
+    }
+  }
+
+  findStudentByID(studentID) {
+    if (studentID === undefined || isNaN(studentID)) {
+      throw new Error('Invalid or missing student ID')
+    }
+    for (const cohort of this.cohortList) {
+      const studentFound = cohort.cohortStudents.find(
+        (item) => item.studentID === studentID
+      )
+      if (studentFound) {
+        const studentSearchResult = {
+          ...studentFound,
+          cohort: cohort.cohortName
+        }
+        return studentSearchResult
+      }
+    }
+    throw new Error('No student matching that ID was found')
```

Comment:
Does throwing the error in this way give the error message if that test fails? Does this work in the same way as returning a "fail condition" in your function would?

Reply:
Throw new Error will display the error message in the console, so if you're running your code through node, or it's the back end code for a live deployment it'll show the message on the console. If you wanted to show an error to the user you'd need to do something else, have an action triggered by the error. That's where something like a "try... catch" comes in.

"try... catch" is essentially "try and run this function, but if an error is thrown, catch it and do this other thing"

In this function I am essentially using this last throw error as a general fail state. The function first checks to see if the given ID is undefined or not a number, if it passes that check then it runs a search for that ID and if that search doesn't find a student then this error is thrown. If a student is found then the function returns and this line isn't reached.

Make sense?

Reply:
Nice one! Thanks for explaining that for me!